<a href="https://colab.research.google.com/github/basunepal/Seizure-Detection/blob/main/Epileptic_Seizure_Recognition_in_Unbalanced_and_Balanced_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
data = pd.read_csv("/content/drive/MyDrive/Epileptic Seizure Recognition.csv")

In [ ]:
data

,Unnamed,X1,X2,X3,X4,X5,X6,X7,X8,X9,...,X170,X171,X172,X173,X174,X175,X176,X177,X178,y
0,X21.V1.791,135,190,229,223,192,125,55,-9,-33,...,-17,-15,-31,-77,-103,-127,-116,-83,-51,4
1,X15.V1.924,386,382,356,331,320,315,307,272,244,...,164,150,146,152,157,156,154,143,129,1
2,X8.V1.1,-32,-39,-47,-37,-32,-36,-57,-73,-85,...,57,64,48,19,-12,-30,-35,-35,-36,5
3,X16.V1.60,-105,-101,-96,-92,-89,-95,-102,-100,-87,...,-82,-81,-80,-77,-85,-77,-72,-69,-65,5
4,X20.V1.54,-9,-65,-98,-102,-78,-48,-16,0,-21,...,4,2,-12,-32,-41,-65,-83,-89,-73,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11495,X22.V1.114,-22,-22,-23,-26,-36,-42,-45,-42,-45,...,15,16,12,5,-1,-18,-37,-47,-48,2
11496,X19.V1.354,-47,-11,28,77,141,211,246,240,193,...,-65,-33,-7,14,27,48,77,117,170,1
11497,X8.V1.28,14,6,-13,-16,10,26,27,-9,4,...,-65,-48,-61,-62,-67,-30,-2,-1,-8,5
11498,X10.V1.932,-40,-25,-9,-12,-2,12,7,19,22,...,121,135,148,143,116,86,68,59,55,3


In [ ]:
data.describe()

,X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,...,X170,X171,X172,X173,X174,X175,X176,X177,X178,y
count,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.00000,11500.00000,11500.000000,...,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000,11500.000000
mean,-11.581391,-10.911565,-10.187130,-9.143043,-8.009739,-7.003478,-6.502087,-6.68713,-6.55800,-6.168435,...,-10.145739,-11.630348,-12.943478,-13.668870,-13.363304,-13.045043,-12.705130,-12.426000,-12.195652,3.000000
std,165.626284,166.059609,163.524317,161.269041,160.998007,161.328725,161.467837,162.11912,162.03336,160.436352,...,164.652883,166.149790,168.554058,168.556486,167.257290,164.241019,162.895832,162.886311,164.852015,1.414275
min,-1839.000000,-1838.000000,-1835.000000,-1845.000000,-1791.000000,-1757.000000,-1832.000000,-1778.00000,-1840.00000,-1867.000000,...,-1867.000000,-1865.000000,-1642.000000,-1723.000000,-1866.000000,-1863.000000,-1781.000000,-1727.000000,-1829.000000,1.000000
25%,-54.000000,-55.000000,-54.000000,-54.000000,-54.000000,-54.000000,-54.000000,-55.00000,-55.00000,-54.000000,...,-55.000000,-56.000000,-56.000000,-56.000000,-55.000000,-56.000000,-55.000000,-55.000000,-55.000000,2.000000
50%,-8.000000,-8.000000,-7.000000,-8.000000,-8.000000,-8.000000,-8.000000,-8.00000,-7.00000,-7.000000,...,-9.000000,-10.000000,-10.000000,-10.000000,-10.000000,-9.000000,-9.000000,-9.000000,-9.000000,3.000000
75%,34.000000,35.000000,36.000000,36.000000,35.000000,36.000000,35.000000,36.00000,36.00000,35.250000,...,34.000000,34.000000,33.000000,33.000000,34.000000,34.000000,34.000000,34.000000,34.000000,4.000000
max,1726.000000,1713.000000,1697.000000,1612.000000,1518.000000,1816.000000,2047.000000,2047.00000,2047.00000,2047.000000,...,1777.000000,1472.000000,1319.000000,1436.000000,1733.000000,1958.000000,2047.000000,2047.000000,1915.000000,5.000000


In [ ]:
data.shape

(11500, 180)

In [ ]:
data['y'].value_counts()

y
4    2300
1    2300
5    2300
2    2300
3    2300
Name: count, dtype: int64

In the data, 1 means having seizure and (2,3,4,5) means having other brain activity. So, we replace (2,3,4,5) with 0.

0 means having no seizure

1 means having seizure.



In [ ]:
data['y'].replace((2,3,4,5),0, inplace=True)

In [ ]:
data['y'].value_counts()

y
0    9200
1    2300
Name: count, dtype: int64

Our data now has 2300 sample of seizure activity and 9200 samples of non-seizure activity.

In [ ]:
data

,Unnamed,X1,X2,X3,X4,X5,X6,X7,X8,X9,...,X170,X171,X172,X173,X174,X175,X176,X177,X178,y
0,X21.V1.791,135,190,229,223,192,125,55,-9,-33,...,-17,-15,-31,-77,-103,-127,-116,-83,-51,0
1,X15.V1.924,386,382,356,331,320,315,307,272,244,...,164,150,146,152,157,156,154,143,129,1
2,X8.V1.1,-32,-39,-47,-37,-32,-36,-57,-73,-85,...,57,64,48,19,-12,-30,-35,-35,-36,0
3,X16.V1.60,-105,-101,-96,-92,-89,-95,-102,-100,-87,...,-82,-81,-80,-77,-85,-77,-72,-69,-65,0
4,X20.V1.54,-9,-65,-98,-102,-78,-48,-16,0,-21,...,4,2,-12,-32,-41,-65,-83,-89,-73,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11495,X22.V1.114,-22,-22,-23,-26,-36,-42,-45,-42,-45,...,15,16,12,5,-1,-18,-37,-47,-48,0
11496,X19.V1.354,-47,-11,28,77,141,211,246,240,193,...,-65,-33,-7,14,27,48,77,117,170,1
11497,X8.V1.28,14,6,-13,-16,10,26,27,-9,4,...,-65,-48,-61,-62,-67,-30,-2,-1,-8,0
11498,X10.V1.932,-40,-25,-9,-12,-2,12,7,19,22,...,121,135,148,143,116,86,68,59,55,0


We seperate the feature and target variable from the data so we create a variable X and y.

X holds the features

y holds the target

In [ ]:
X = data.drop(['Unnamed','y'],axis=1)
X

,X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,...,X169,X170,X171,X172,X173,X174,X175,X176,X177,X178
0,135,190,229,223,192,125,55,-9,-33,-38,...,8,-17,-15,-31,-77,-103,-127,-116,-83,-51
1,386,382,356,331,320,315,307,272,244,232,...,168,164,150,146,152,157,156,154,143,129
2,-32,-39,-47,-37,-32,-36,-57,-73,-85,-94,...,29,57,64,48,19,-12,-30,-35,-35,-36
3,-105,-101,-96,-92,-89,-95,-102,-100,-87,-79,...,-80,-82,-81,-80,-77,-85,-77,-72,-69,-65
4,-9,-65,-98,-102,-78,-48,-16,0,-21,-59,...,10,4,2,-12,-32,-41,-65,-83,-89,-73
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11495,-22,-22,-23,-26,-36,-42,-45,-42,-45,-49,...,20,15,16,12,5,-1,-18,-37,-47,-48
11496,-47,-11,28,77,141,211,246,240,193,136,...,-94,-65,-33,-7,14,27,48,77,117,170
11497,14,6,-13,-16,10,26,27,-9,4,14,...,-42,-65,-48,-61,-62,-67,-30,-2,-1,-8
11498,-40,-25,-9,-12,-2,12,7,19,22,29,...,114,121,135,148,143,116,86,68,59,55


In [ ]:
X.shape

(11500, 178)

In [ ]:
y = data['y']
y

0        0
1        1
2        0
3        0
4        0
        ..
11495    0
11496    1
11497    0
11498    0
11499    0
Name: y, Length: 11500, dtype: int64

In [ ]:
y.shape

(11500,)

In [ ]:
y.value_counts()

y
0    9200
1    2300
Name: count, dtype: int64

#Working on Imbalanced Data

Target variable y has imbalanced dataset

2300 Seizure samples
9200 Non Seizure samples

## Train Test Split

Now we split the X and y into test and training data to fit them to model

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=42)

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((8050, 178), (3450, 178), (8050,), (3450,))

After splitting,

8050 Training data
3450 Test data

##Model Building

## Logistic Regression

We import the model and initialize it

In [ ]:
from sklearn.linear_model import LogisticRegression
lg = LogisticRegression(solver='lbfgs', max_iter=10000)

Fit the training data

In [ ]:
lg.fit(X_train, y_train)

LogisticRegression(max_iter=10000)

In [ ]:
y_lg = lg.predict(X_test)

###Model Evaluation

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_lg))

              precision    recall  f1-score   support

           0       0.82      0.99      0.90      2752
           1       0.87      0.14      0.24       698

    accuracy                           0.82      3450
   macro avg       0.85      0.57      0.57      3450
weighted avg       0.83      0.82      0.77      3450



In [ ]:
from sklearn import metrics

lg_accuracy = metrics.accuracy_score(y_test, y_lg)
lg_precision = metrics.precision_score(y_test, y_lg)
lg_recall = metrics.recall_score(y_test, y_lg)
lg_metrics = (lg_accuracy, lg_precision, lg_recall)
lg_metrics

(0.8217391304347826, 0.8738738738738738, 0.13896848137535817)

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_lg)

array([[2738,   14],
       [ 601,   97]])

##Random Forest

We import and initialize the model

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()

Fitting the training data

In [ ]:
rf.fit(X_train, y_train)

In [ ]:
y_rf = rf.predict(X_test)

###Model Evaluation

In [ ]:
print(classification_report(y_test,y_rf))

In [ ]:
rf_accuracy = metrics.accuracy_score(y_test, y_rf)
rf_precision = metrics.precision_score(y_test, y_rf)
rf_recall = metrics.recall_score(y_test, y_rf)
rf_metrics = (rf_accuracy, rf_precision, rf_recall)
rf_metrics

In [ ]:
confusion_matrix(y_test, y_rf)

##XGBClassifier

We import the model and initialize it

In [ ]:
from xgboost import XGBClassifier
xgbc = XGBClassifier()

Fit the training data

In [ ]:
xgbc.fit(X_train, y_train)

In [ ]:
y_xgbc = xgbc.predict(X_test)

###Model Evaluation

In [ ]:
print(classification_report(y_test,y_xgbc))

In [ ]:
xgbc_accuracy = metrics.accuracy_score(y_test, y_xgbc)
xgbc_precision = metrics.precision_score(y_test, y_xgbc)
xgbc_recall = metrics.recall_score(y_test, y_xgbc)
xgbc_metrics = (xgbc_accuracy, xgbc_precision, xgbc_recall)
xgbc_metrics

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_xgbc)

### Parameter tuning

We find the best parameter for the XGBC Classifier usign Gridsearch and evaluate the model

In [ ]:
xgbc_tuned = XGBClassifier(learning_rate=0.2, max_depth=10, min_child_weight=5, n_estimators=125)

In [ ]:
xgbc_tuned.fit(X_train, y_train)

In [ ]:
y_xgbc_tuned = xgbc_tuned.predict(X_test)

In [ ]:
print(classification_report(y_test,y_xgbc_tuned))

In [ ]:
xgbc_tuned_accuracy = metrics.accuracy_score(y_test, y_xgbc_tuned)
xgbc_tuned_precision = metrics.precision_score(y_test, y_xgbc_tuned)
xgbc_tuned_recall = metrics.recall_score(y_test, y_xgbc_tuned)
xgbc_tuned_metrics = (xgbc_accuracy, xgbc_precision, xgbc_recall)
xgbc_tuned_metrics

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_xgbc_tuned)

###GridSearchCV

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
cv_params = {'max_depth': [1,4,10],
             'min_child_weight': [1,3,5],
             'learning_rate': [0.1, 0.2, 0.3],
             'n_estimators': [75, 100, 125]
             }

scoring = {'accuracy', 'precision', 'recall', 'f1'}

xgbc_cv = GridSearchCV(xgbc, cv_params, scoring=scoring, cv=5, refit='recall')

In [ ]:
%%time
#xgbc_cv.fit(X_train, y_train)

In [ ]:
#xgbc_cv.best_params_

#Working on Balanced Data

##SMOTE

After separating features and target,

X has 11500 rows and 178 features

y has 11500 rows where 2300 are seizure data and 9200 are non seizure data

Now we use the SMOTE technique to balanced the data. We need to generate 9200 - 2300 = 6900 new data using the SMOTE technique.



Import the SMOTE library and initialize it

In [ ]:
from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state=42)

In [ ]:
X_bal, y_bal = sm.fit_resample(X, y)

Check the size of feature variable before and after SMOTE

In [ ]:
X.shape, X_bal.shape

Check the size of target variable before and after SMOTE

In [ ]:
y.shape, y_bal.shape

In [ ]:
y_bal.value_counts()

Before SMOTE,

X had 11500 rows
y had 11500 rows
y had 2300 are seizure data and 9200 are non seizure data

After SMOTE,

X_bal has 18400 rows
y_bal has 18400 rows
y_bal has 9200 seizure data and 9200 non seizure data

##Train Test Split

Now we split the X_bal and y_bal into test and training data to fit them to model

In [ ]:
from sklearn.model_selection import train_test_split
X_bal_train, X_bal_test, y_bal_train, y_bal_test = train_test_split(X_bal,y_bal,test_size=0.3,random_state=42)

In [ ]:
X_bal_train.shape, X_bal_test.shape, y_bal_train.shape, y_bal_test.shape

After splitting,

12880 Training data
5520 Test data

## Model Building




## Logistic Regression

We import the model and initialize it

In [ ]:
from sklearn.linear_model import LogisticRegression
lg_bal = LogisticRegression(solver='lbfgs', max_iter=10000)

Fit the training data

In [ ]:
lg_bal.fit(X_bal_train,y_bal_train)

In [ ]:
y_bal_lg = lg_bal.predict(X_bal_test)

###Model Evaluation

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_bal_test,y_bal_lg))

In [ ]:
from sklearn import metrics

lg_bal_accuracy = metrics.accuracy_score(y_bal_test, y_bal_lg)
lg_bal_precision = metrics.precision_score(y_bal_test, y_bal_lg)
lg_bal_recall = metrics.recall_score(y_bal_test, y_bal_lg)
lg_bal_metrics = (lg_bal_accuracy, lg_bal_precision, lg_bal_recall)
lg_bal_metrics

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_bal_test, y_bal_lg)

In [ ]:
lg_metrics, lg_bal_metrics

##Random Forest

We import and initialize the model

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_bal = RandomForestClassifier()

Fitting the training data

In [ ]:
rf_bal.fit(X_bal_train, y_bal_train)

RandomForestClassifier()

In [ ]:
y_bal_rf = rf_bal.predict(X_bal_test)

###Model Evaluation

In [ ]:
print(classification_report(y_bal_test,y_bal_rf))

              precision    recall  f1-score   support

           0       0.99      0.97      0.98      2778
           1       0.97      0.99      0.98      2742

    accuracy                           0.98      5520
   macro avg       0.98      0.98      0.98      5520
weighted avg       0.98      0.98      0.98      5520



In [ ]:
rf_bal_accuracy = metrics.accuracy_score(y_bal_test, y_bal_rf)
rf_bal_precision = metrics.precision_score(y_bal_test, y_bal_rf)
rf_bal_recall = metrics.recall_score(y_bal_test, y_bal_rf)
rf_bal_metrics = (rf_bal_accuracy, rf_bal_precision, rf_bal_recall)
rf_bal_metrics

(0.9766304347826087, 0.9677765843179377, 0.9857768052516411)

In [ ]:
confusion_matrix(y_bal_test, y_bal_rf)

array([[2688,   90],
       [  39, 2703]])

In [ ]:
rf_metrics, rf_bal_metrics

((0.972463768115942, 0.935064935064935, 0.9283667621776505),
 (0.9766304347826087, 0.9677765843179377, 0.9857768052516411))

##XGBClassifier

We import the model and initialize it

In [ ]:
from xgboost import XGBClassifier
xgbc_bal = XGBClassifier()

Fit the training data

In [ ]:
xgbc_bal.fit(X_bal_train, y_bal_train)

In [ ]:
y_bal_xgbc = xgbc.predict(X_bal_test)

###Model Evaluation

In [ ]:
print(classification_report(y_bal_test,y_bal_xgbc))

In [ ]:
xgbc_bal_accuracy = metrics.accuracy_score(y_bal_test, y_bal_xgbc)
xgbc_bal_precision = metrics.precision_score(y_bal_test, y_bal_xgbc)
xgbc_bal_recall = metrics.recall_score(y_bal_test, y_bal_xgbc)
xgbc_bal_metrics = (xgbc_bal_accuracy, xgbc_bal_precision, xgbc_bal_recall)
xgbc_bal_metrics

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_bal_test, y_bal_xgbc)

### Parameter tuning

We find the best parameter for the XGBC Classifier usign Gridsearch and evaluate the model

In [ ]:
xgbc_bal_tuned = XGBClassifier(learning_rate=0.1, max_depth=25, min_child_weight=5, n_estimators=150)

In [ ]:
xgbc_bal_tuned.fit(X_bal_train, y_bal_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=25, max_leaves=None,
              min_child_weight=5, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=150, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

In [ ]:
y_bal_xgbc_tuned = xgbc_bal_tuned.predict(X_bal_test)

In [ ]:
print(classification_report(y_bal_test,y_bal_xgbc_tuned))

              precision    recall  f1-score   support

           0       0.99      0.98      0.98      2778
           1       0.98      0.99      0.98      2742

    accuracy                           0.98      5520
   macro avg       0.98      0.98      0.98      5520
weighted avg       0.98      0.98      0.98      5520



In [ ]:
xgbc_bal_tuned_accuracy = metrics.accuracy_score(y_bal_test, y_bal_xgbc_tuned)
xgbc_bal_tuned_precision = metrics.precision_score(y_bal_test, y_bal_xgbc_tuned)
xgbc_bal_tuned_recall = metrics.recall_score(y_bal_test, y_bal_xgbc_tuned)
xgbc_bal_tuned_metrics = (xgbc_accuracy, xgbc_precision, xgbc_recall)
xgbc_bal_tuned_metrics

(0.9707246376811595, 0.954337899543379, 0.8982808022922636)

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_bal_test, y_bal_xgbc_tuned)

array([[2712,   66],
       [  41, 2701]])

In [ ]:
xgbc_metrics, xgbc_tuned_metrics, xgbc_bal_metrics, xgbc_bal_tuned_metrics

((0.9707246376811595, 0.954337899543379, 0.8982808022922636),
 (0.9707246376811595, 0.954337899543379, 0.8982808022922636),
 (0.9471014492753623, 0.9951495553759094, 0.8978847556528081),
 (0.9707246376811595, 0.954337899543379, 0.8982808022922636))

###GridSearchCV

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
cv_params = {'max_depth': [1,4,10],
             'min_child_weight': [1,3,5],
             'learning_rate': [0.1, 0.2, 0.3],
             'n_estimators': [75, 100, 125]
             }

scoring = {'accuracy', 'precision', 'recall', 'f1'}

xgbc_cv = GridSearchCV(xgbc, cv_params, scoring=scoring, cv=5, refit='recall')

In [ ]:
%%time
#xgbc.fit(X_bal_train, y_bal_train)

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 7.87 µs


In [ ]:
#xgbc_cv.best_params_

# Classification of Balanced Dataset

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_sm_pred))

NameError: name 'y_sm_pred' is not defined

In [ ]:
confusion_matrix(y_test, y_pred)

Balanced Data set has improved the recall that means we prodicted more instanced of seizure than the unbalanced dataset


In [ ]:
confusion_matrix(y_test, y_sm_pred)

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf = RandomForestClassifier()
rf.fit(X_sm_train, y_sm_train)

In [ ]:
y_sm_pred = rf.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_sm_pred))